pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+                    }
+                    err_info
+                })
+        }
+
+        /// Change the instantiation fee of the smart extension template
+        ///
+        /// # Arguments
+        /// * origin - Only owner of template is allowed to execute the dispatchable.
+        /// * code_hash - Unique hash of the smart extension template.
+        /// * new_instantiation_fee - New value of instantiation fee to the smart extension template.
+        #[weight = 1000_000_000]
+        pub fn change_instantiation_fee(origin, code_hash: CodeHash<T>, new_instantiation_fee: BalanceOf<T>) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(<TemplateMetaDetails<T>>::contains_key(code_hash), Error::<T>::TemplateNotExists);
+            Self::deposit_event(RawEvent::InstantiationFeeChanged(did, code_hash, Self::get_template_meta_details(code_hash).meta_info.instantiation_fee, new_instantiation_fee));
+            <TemplateMetaDetails<T>>::mutate(&code_hash, |meta_details| meta_details.meta_info.instantiation_fee = new_instantiation_fee);
+            Ok(())
+        }
+
+        /// Allows a smart extension template owner to freeze the instantiation.
+        ///
+        /// # Arguments
+        /// * origin - Only owner of the template is allowed to execute the dispatchable.
+        /// * code_hash - Unique hash of the smart extension template.
+        #[weight = 1_000_000_000]
+        pub fn freeze_instantiation(origin, code_hash: CodeHash<T>) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(<TemplateMetaDetails<T>>::contains_key(code_hash), Error::<T>::TemplateNotExists);
+             // Access the meta details of SE template
+            let meta_details = Self::get_template_meta_details(code_hash);
+            ensure!(!meta_details.is_instantiation_freezed(), Error::<T>::InstantiationAlreadyFreezed);
+            <TemplateMetaDetails<T>>::mutate(&code_hash, |meta_details| meta_details.is_freeze = true);
+            Self::deposit_event(RawEvent::InstantiationFreezed(did, code_hash));
+            Ok(())
+        }
+
+        /// Allows a smart extension template owner to un freeze the instantiation.
+        ///
+        /// # Arguments
+        /// * origin - Only owner of the template is allowed to execute the dispatchable.
+        /// * code_hash - Unique hash of the smart extension template.
+        #[weight = 1_000_000_000]
+        pub fn unfreeze_instantiation(origin, code_hash: CodeHash<T>) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(<TemplateMetaDetails<T>>::contains_key(code_hash), Error::<T>::TemplateNotExists);
```

Comment:
Not sure if this will be relevant when not WIP, but 3 first lines in these functions could be extracted.